src/ConfigurationClientManager.ts

```diff
@@ -0,0 +1,310 @@
+        if (this.isFailoverable &&
+            currentTime > new Date(this.#lastFallbackClientRefreshAttempt + MINIMAL_CLIENT_REFRESH_INTERVAL).getTime()) {
+            this.#lastFallbackClientRefreshAttempt = currentTime;
+            const host = new URL(this.endpoint).hostname;
+            await this.#discoverFallbackClients(host);
+        }
+    }
+
+    async #discoverFallbackClients(host) {
+        const timeout = setTimeout(() => {
+        }, SRV_QUERY_TIMEOUT);
+        const srvResults = await querySrvTargetHost(host);
+
+        try {
+            const result = await Promise.race([srvResults, timeout]);
+
+            if (result === timeout) {
+                throw new Error("SRV record query timed out.");
+            }
+
+            const srvTargetHosts = result as string[];
+            // Shuffle the list of SRV target hosts
+            for (let i = srvTargetHosts.length - 1; i > 0; i--) {
+                const j = Math.floor(Math.random() * (i + 1));
+                [srvTargetHosts[i], srvTargetHosts[j]] = [srvTargetHosts[j], srvTargetHosts[i]];
+            }
+
+            const newDynamicClients: ConfigurationClientWrapper[] = [];
+            for (const host of srvTargetHosts) {
+                if (isValidEndpoint(host, this.#validDomain)) {
+                    const targetEndpoint = `https://${host}`;
+                    if (targetEndpoint.toLowerCase() === this.endpoint.toLowerCase()) {
+                        continue;
+                    }
+                    const client = this.#newConfigurationClient(targetEndpoint);
+                    newDynamicClients.push(new ConfigurationClientWrapper(targetEndpoint, client));
+                }
+            }
+
+            this.#dynamicClients = newDynamicClients;
+            this.#lastFallbackClientRefreshTime = Date.now();
+        } catch (err) {
+            console.warn(`Fail to build fallback clients, ${err.message}`);
+        }
+    }
+
+    #newConfigurationClient(endpoint) {
+        if (this.#credential) {
+            return new AppConfigurationClient(endpoint, this.#credential, this.#clientOptions);
+        }
+
+        const connectionStr = buildConnectionString(endpoint, this.#secret, this.#id);
+        return new AppConfigurationClient(connectionStr, this.#clientOptions);
+    }
+
+    #isFallbackClientDiscoveryDue(dateTime) {
+        return dateTime >= this.#lastFallbackClientRefreshAttempt + MINIMAL_CLIENT_REFRESH_INTERVAL
+            && (!this.#dynamicClients
+                || this.#dynamicClients.every(client => dateTime < client.backoffEndTime)
+                || dateTime >= this.#lastFallbackClientRefreshTime + FALLBACK_CLIENT_REFRESH_EXPIRE_INTERVAL);
+    }
+}
+
+/**
+ * Query SRV records and return target hosts.
+ */
+async function querySrvTargetHost(host: string): Promise<string[]> {
+    const results: string[] = [];
+    let dns;
+
+    if (typeof global !== "undefined" && global.dns) {
+        dns = global.dns;
```

Comment:
I just found that we cannot use global.dns in script enviroment. It only works in Node.js shell environment.
We have to write something like:
``` js
const dns = await import("dns");
```

And in this case, it should be a part of the process determining whether current environment supports replica auto discovery. We should only import for once and the loaded module should be stored as a property of the client manager.